mmrotate/core/bbox/coder/angle_coder.py

```diff
@@ -0,0 +1,114 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import math
+
+import torch
+from mmdet.core.bbox.coder.base_bbox_coder import BaseBBoxCoder
+
+from ..builder import ROTATED_BBOX_CODERS
+
+
+@ROTATED_BBOX_CODERS.register_module()
+class CSLCoder(BaseBBoxCoder):
+    """Circular Smooth Label Coder.
+
+    `Circular Smooth Label (CSL)
+    <https://link.springer.com/chapter/10.1007/978-3-030-58598-3_40>`_ .
+
+    Args:
+        angle_version (str): Angle definition.
+        omega (float, optional): Angle discretization granularity.
+            Default: 1.
+        window (str, optional): Window function. Default: gaussian.
+        radius (int/float): window radius, int type for
+            ['triangle', 'rect', 'pulse'], float type for
+            ['gaussian']. Default: 6.
+    """
+
+    def __init__(self, angle_version, omega=1, window='gaussian', radius=6):
+        super().__init__()
+        self.angle_version = angle_version
+        assert angle_version in ['oc', 'le90', 'le135']
+        assert window in ['gaussian', 'triangle', 'rect', 'pulse']
+        self.angle_range = 90 if angle_version == 'oc' else 180
+        self.angle_offset_dict = {'oc': 0, 'le90': 90, 'le135': 45}
+        self.angle_offset = self.angle_offset_dict[angle_version]
+        self.omega = omega
+        self.window = window
+        self.radius = radius
+        self.coding_len = int(self.angle_range // omega)
+
+    def encode(self, angle_targets):
+        """Circular Smooth Label Encoder.
+
+        Args:
+            angle_targets (Tensor): Angle offset for each scale level
+                Has shape (num_anchors * H * W, 1)
+
+        Returns:
+            list[Tensor]: The csl encoding of angle offset for each
+                scale level. Has shape (num_anchors * H * W, coding_len)
+        """
+
+        # radius to degree
+        angle_targets_deg = angle_targets * (180 / math.pi)
+        # empty label
+        smooth_label = torch.zeros_like(angle_targets).repeat(
+            1, self.coding_len)
+        angle_targets_deg = (angle_targets_deg +
+                             self.angle_offset) / self.omega
+        # Float to Int
+        angle_targets_long = angle_targets_deg.long()
+
+        if self.window == 'pulse':
+            radius_range = angle_targets_long
```

Comment:
```suggestion
            radius_range = angle_targets_long % self.coding_len
```
an error occurred when use omega=8 window=pulse

Reply:
done